### ISSUE TITLE
Change API to return a stream with a hash Promise property.

### ISSUE THREAD

**Issue by pgte** [aickin/react-dom-stream#2]
Instead of passing the destination stream into the function call, why not returning a readable stream that you can pipe?

**Comment by tadjik1** [comment]
@pgte do you mean something like (in koajs)?

```javascript
app.use(function* applicationMiddleware() {
  this.body = ReactDOMStream.renderToString(<App />);
});
```

**Comment by aickin** [comment]
@pgte Thanks for the feedback; it's a great question.

I thought about having it return a stream and decided against it, but I may well be wrong. The core reason is that the `renderToString` function also has to asynchronously return a checksum, which is used to render on the client side. I figured that the current signature would be the easiest for folks migrating from `ReactDOM.renderToString`. Being honest, I am also biased by the fact that I have about 100x more experience with Promises than with Streams.

If we had `renderToString` return a stream, I think that the server example in the doc would change to something like this (assuming ES2015/7):

```javascript
// OPTION 1
app.get('/', function (req, res) {
    const stream = ReactDOMStream.renderToString(<Foo prop={value}/>);
    stream.pipe(res, {end:false}); // I think this has to be a separate statement because we need stream further down.
    stream.on("end", () => {
        const hash = stream.getHash(); // or maybe stream.hash?
        // TODO: write the hash out to the page in a script tag
        res.end();
    })
});
```

Or, we could return a structured object from `renderToString` which has both a `stream` and a `hashPromise` member:

```javascript
// OPTION 2
app.get('/', async function (req, res) {
    const {stream, hashPromise} = ReactDOMStream.renderToString(<Foo prop={value}/>);
    stream.pipe(res, {end:false});
    const hash = await hashPromise;
    // TODO: write the hash out to the page in a script tag
    res.end();
});
```

Both of these feel like a bit more boilerplate than I'd like. 

Option 1 feels a little goofy, with the tacked on `getHash` method, but that's probably a pretty arbitrary feeling. I worry that it's a bit more work for a dev to port current code than the existing API, but I may be wrong. I also worry that a dev who is used to streams will want to just add `.pipe(res)` at the end of the function call out of habit, and then they won't be able to get the hash.

Option 2 has the extra problem that it doesn't make any sense to have `renderToStaticMarkup` return an object with a stream property, since `renderToStaticMarkup` doesn't need to return a hash. It would be super awkward to have one method (`renderToStaticMarkup`) return a stream while the other (`renderToStaticMarkup`) returns an object with a stream property.

A few questions for you:
1) Thoughts on these options?
2) Is there another way I'm not thinking of to structure the API to return both a stream and an asynchronous hash? I fully admit that streams are not my strong suit.
3) I could maybe provide methods that implement Option 1 with different function names; would that satisfy you?

**Comment by jakearchibald** [comment]
I don't know much about React, so I'm coming at this from a streams/promise perspective

```js
const stream = ReactDOMStream.renderToString(<Foo prop={value}/>);
stream.hash.then(hashVal => {
  // ...
});
```

The `hash` promise is a property rather than a function, as there's no side-effect from requesting it.

The code that gathers the hash is passive, as in if the stream is never read, `hash` would never resolve. This also means `hash` wouldn't require a buffer of the output string. This could be thought of as:

```
streamingReactToString --pipeTo--> hashGenerator --pipeTo--> httpResponse
```

…where `hashGenerator` is what's returned by `renderToString`. It's a transform stream that simply returns the same value, but it's observing the value and generating the hash.

**Comment by tadjik1** [comment]
Please explain me why you need in checksum on client side?

**Comment by aickin** [comment]
@jakearchibald Thanks for contributing!

I'm trying to understand what you are advocating for, though. The code sample has `hash` as a Promise, but the text below (maybe?) proposes that it be a transform stream. Or are you saying that `hash` in your code sample is *conceptually* like a transform stream, but that in the API, it should actually be a promise?

**Comment by aickin** [comment]
@tadjik1 Thanks for the question!

It's sort of a long story, but the basic answer is that React wants to make sure that the markup generated on the client is the same as the markup generated on the server. If that is the case, it doesn't have to do any DOM manipulation at all when the client render happens.

**Comment by aickin** [comment]
I also just found a potentially thorny problem with a pure stream-based approach. I believe that when gzip is applied with a library like `compression`, I need to explicitly call `flush` to make sure that the content continues to stream even while being buffered somewhat for compression. However, `flush` is not a standard stream method. If I'm right about that (and I'm still investigating), then a purely stream-based solution on Express with `compression` would stop streaming if the developer turns on gzip.

**Comment by jakearchibald** [comment]
Sorry, I've done a bad job at explaining. `stream.hash` is just a promise for the hash. Eg:

```js
function renderToString() {
  const md5 = crypto.createHash('md5');
  const stream = _getReactStringStream().pipe(through2(function(chunk, enc, callback) {
    md5.update(chunk);
    this.push(chunk);
    callback();
  }));

  stream.hash = new Promise((resolve, reject) => {
    stream.on('close', _ => resolve(md5.digest('hex')));
    stream.on('error', _ => reject());
  });

  return stream;
}
```

**Comment by jakearchibald** [comment]
@aickin you don't need to explicitly call flush for compressed streams, although I got better automatic flushing using https://github.com/jakearchibald/offline-wikipedia/blob/master/index.js#L64

**Comment by aickin** [comment]
@jakearchibald Thanks. Also, ha! I've spent the last several hours trying to get compression playing nice with streaming, and that's the one zlib option that isn't mentioned in the `compression` options. I'll try it tomorrow.

By the way, I did find that if I yielded to the event loop with `setImmediate`, compression and streaming seemed to get along, but I'll do some more testing tomorrow.

**Comment by aickin** [comment]
@jakearchibald Got it re: your hash proposal. The internals of how the hash is implemented aren't super important to this discussion, I think. It's enough to say that it's a rolling hash that doesn't buffer the whole stream and requires the render to be finished before it's ready.

So if I'm reading you correctly, your proposal is only slightly different than the current API. 

Current API:
```javascript
const hash = ReactDOMStream.renderToString(<Foo prop={value}/>, res);
hash.then(hashVal => {
  // TODO: write hash out to page.
  res.end()
});
```
Your proposal:
```javascript
const stream = ReactDOMStream.renderToString(<Foo prop={value}/>);
stream.pipe(res);
stream.hash.then(hashVal => {
  // TODO: write hash out to page.
  res.end();
});
```
They both have Promises of a hash. The proposal is more flexible if you want to do creative things with the stream, but a tiny bit more awkward to use in the base case. Is that a fair assessment?

And as to what the hash is for, the developer has to read it into the page so that it's available to JS running in the browser.

**Comment by jakearchibald** [comment]
> And as to what the hash is for, the developer has to read it into the page so that it's available to JS running in the browser.

This is the bit that bothers me, eg:

```js
app.get('/', (req, res) => {
  ReactDOMStream.renderToString(<Foo prop={value}/>).pipe(res);
});
```

…how do I get the hash into the page? Or would you use this on top of a streaming template system such as dustjs?

```js
app.get('/', (req, res) => {
  const stream = ReactDOMStream.renderToString(<Foo prop={value}/>);
  dust.stream('home', {
    body: stream,
    hash: stream.hash
  });
});
```

…where the home template is something like…

```js
<html>
<body>
{body}
<script>var expectedHash = "{hash}";</script>
</body>
</html>
```

**Comment by th0r** [comment]
@jakearchibald I've split my page template (I use `nunjucks`) into two parts: `pageHeader` and `pageFooter` and use it like so:
```js
// Setting proper content type (in other case `compression` middleware won't gzip response)
res.type('html');

// Rendering and sending page header
res.write(
    views.render('pageHeader.html', {
        title: DocumentMetaData.title
    })
);
// Flushing header here for browser to begin downloading assets (styles, scripts etc.)
res.flush();

// Rendering and sending page content (react part)
const markupHash = await ReactDomStream.renderToString(<Page/>, res);

// Rendering and sending page footer
res.end(
    views.render('pageFooter.html', { markupHash })
);
```
`pageHeader.html`:
```jinja
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>{{ title }}</title>
    <meta name="viewport" content="width=device-width, user-scalable=no"/>

    <link rel="stylesheet" href="{{ assetPath('styles/app.vendor.css') }}"/>
    <link rel="stylesheet" href="{{ assetPath('styles/app.css') }}"/>

    <script src="{{ scriptPath('app') }}" defer></script>
</head>
<body>
    <div id="app">{# React DOM Stream will render from here #}
```
`pageFooter.html`:
```jinja
{# React DOM Stream have rendered app markup here #}</div>
    <script>
        var App = {
            markupHash: {{ markupHash | js }}
        };
    </script>
</body>
</html>
```

**Comment by jakearchibald** [comment]
Ahh ok, so if it returned a stream it'd be:

```js
// Rendering and sending page header
res.write(
    views.render('pageHeader.html', {
        title: DocumentMetaData.title
    })
);

const stream = ReactDomStream.renderToString(<Page/>);
stream.pipe(res);

// Rendering and sending page footer
res.end(
    views.render('pageFooter.html', { await stream.hash })
);
```

…which isn't quite as neat I guess, but the current system makes using dustjs impossible, returning a stream makes it possible.

**Comment by th0r** [comment]
@jakearchibald Well, I think it won't change a lot:
```js
// ...

const stream = ReactDomStream.renderToString(<Page/>);
stream.pipe(res);

const markupHash = await stream.hash;

// Rendering and sending page footer
res.end(
    views.render('pageFooter.html', { markupHash })
);
```
The only difference I see is adding `stream.pipe(res)` which is totally ok as for me.

**Comment by aickin** [comment]
Ah, I see the root of the confusion (as it were). Yeah, generally you shouldn't use React to render an entire HTML page on the client. In fact, [you shouldn't even use React to render the entire `body` element](https://medium.com/@dan_abramov/two-weird-tricks-that-fix-react-7cf9bbdef375#.f4jisbpf8), because other libraries tend to add children to the body and confuse React, leading to bugs very difficult to reproduce and fix. 

It also is dangerous to use React to render the `head` of the document and reconnect on the client; if you mess up the checksum, you could end up rewriting the entire head, which can make script order load inconsistent or even run scripts twice.

The best practice on the client side is to render into one or more `div`s underneath the `body` tag, so that's what we do when we are rendering server side.

In projects I've worked on, we haven't bothered with whole page templates, we just called something like `res.write("<html><head>....</head><body><div id='root'>");` before the React render and res.write("...</div></body></html>") after. Templates would be nice, and I think are a great argument for returning a stream.

One small nitpick on the above code samples: since React is never going to render the entire page, you will always have to call pipe with `{end: false}`, I think. A little annoying, but not too bad.

**Comment by jakearchibald** [comment]
Good point about `{end: false}` - another thing a streaming template engine will sort for you :smile:

**Comment by aickin** [comment]
All right, all right, you've convinced me. ;) I'll put in on the list for v0.2. Thanks for a respectful discussion!

It occurs to me that this could help with the problems encountered in issue #4. In that issue, the bug reporter was encountering an error with a framework that did something like this (with standard React, not react-dom-stream):

```javascript
// render the dynamic code of the page to a string.
var appHtml = ReactDOM.renderToString(<Application/>);

// now use React as a static templating language to create the <html>, 
// <head>, and <body> tags
var fullPageHtml = ReactDOM.renderToStaticMarkup(
    <html>
        <head>
        </head>
        <body>
            <div id="root">{appHtml}</div>
        </body>
    </html>
);
res.send(fullPageHtml);
```

It's OK to render `<html>`, `<body>`, and `<head>` tags using `renderToStaticMarkup` (as opposed to `renderToString`), because `rTSM` creates markup that cannot be reconnected to React on the client side.

With the current react-dom-stream API, it's basically impossible to make this kind of "renders inside of renders" strategy play nice with streaming. However, if I implement @jakearchibald's suggestion, I could do something like this:

```javascript
// render the dynamic code of the page to a stream.
var appHtmlStream = ReactDOMStream.renderToString(<Application/>);

// now use React as a static templating language to create the <html>, 
// <head>, and <body> tags
ReactDOMStream.renderToStaticMarkup(
    <html>
        <head>
        </head>
        <body>
            <div id="root">{appHtmlStream}</div>
            <script>
                window.hash = {appHtmlStream.hash};
            </script>
        </body>
    </html>
).pipe(res); // note, no {end:false} because rTSM generates the entire page.
```

This would also require that `renderToStaticMarkup` have first class support for streams and Promises as children of React elements the way it currently has first class support for strings. For streams, `rTSM` should hand over control of the output to the stream, and for Promises (or thenables), it should wait for result before continuing.

This would allow folks to server-side template completely in React, rather than having to load a separate templating library like dust for the page html. 

I need to think through the implications of this, but at first blush it looks pretty neat; stream all the things! Given that there are other solutions (like dust or just directly calling `res.write`), it's a little down the priority stack for me, but it's still a good idea.

Also worth noting that I wouldn't support nesting of streams and promises in `renderToString`, though, as supporting that on the client side would be kind of a nightmare.

Changing the title of the topic to reflect the desire to change to a stream-returning API.

**Comment by jakearchibald** [comment]
> Thanks for a respectful discussion!

A respectful and productive discussion on the internet? We should have this framed.

> supporting [streams] on the client side would be kind of a nightmare.

There's no access to the browser's streaming parser right now, so there's no point in looking at the page side of things yet.

The ability for the web to stream is one of its real advantages, and it's pretty sad that most serverside frameworks don't support it. Projects like this are a huge step forward.

**Comment by aickin** [comment]
Update: Inspired by this issue, I rewrote the react fork from the ground up with streams behaving as they are supposed to, and I was hoping to release it today, but I ran into a really sticky problem. It seems that piping a Readable to the `res` output stream in Express is several times slower than the strategy I was using for v0.1.x: batching writes and passing through to `res.write`. 

I've written a simple test case that doesn't use any of my react-dom-stream code, and batched writes to `res` beat a piped Readable stream by 614% in a server siege. I'm a bit stuck and would appreciate any help from folks on this thread who are more knowledgeable than I am. I'm also packaging my test case so that I can send it around and hopefully get some feedback on it.

**Comment by aickin** [comment]
I just released version 0.2.0 to npm, which has a Stream-based API. It turned out to be a lot more work than I thought it would be to implement a Stream API that could be competitive with the String API, but I'm happy it's done.

Also: happily, I figured out how to no longer need the hash return value, so the API now just returns a plain old stream.Readable.

Thanks to @pgte and @jakearchibald for engaging on this issue!
